users/views.py

```diff
@@ -440,21 +447,136 @@ def change_username_api_view(request: Request):
 
 
 @api_view(["PATCH"])
-def update_profile_api_view(request: Request):
-    user = request.user
+def update_profile_api_view(request: Request) -> Response:
+    user: User = request.user
     serializer = UserUpdateProfileSerializer(user, data=request.data, partial=True)
     serializer.is_valid(raise_exception=True)
 
+    assert isinstance(request.data, dict)
+    is_spam, reasoning = _check_for_spam(user, request.data)
+
+    if is_spam:
+        return _handle_spam_detection(
+            user, _combine_bio_and_websites_from_request_data(request.data), reasoning
+        )
+    assert isinstance(serializer, UserUpdateProfileSerializer)
+    return _update_user_profile(user, serializer)
+
+
+def _combine_bio_and_websites_from_request_data(request_data: dict) -> str:
+    bio: str | None = request_data.get("bio")
+    website: str | None = request_data.get("website")
+    if not bio:
+        return ""
+    bio_plus_websites = bio + (f"\n\nWebsite: {website}" if website else "")
+    return bio_plus_websites
+
+
+def _check_for_spam(user: User, request_data: dict) -> tuple[bool, str]:  # NOSONAR
+    days_since_joined = (timezone.now() - user.date_joined).days
+    days_since_joined_threshold = 7
+    bio_plus_websites = _combine_bio_and_websites_from_request_data(request_data)
+
+    if not bio_plus_websites:
+        return False, "No bio to check for spam"
+
+    if len(bio_plus_websites) < 10:
+        return False, "Bio is too short to be spam"
+
+    if days_since_joined > days_since_joined_threshold:
+        return (
+            False,
+            f"The user has been a member for more than {days_since_joined_threshold} days",
+        )
+
+    if len(bio_plus_websites) > 25000:
+        return True, "Bio is more than 25000 characters"
+
+    return asyncio.run(_ask_gpt_to_check_for_spam(bio_plus_websites, user.email))
+
+
+def _handle_spam_detection(
+    user: User, bio_plus_websites: str, reasoning: str
+) -> Response:
+    user.soft_delete()
+    user.save()
+    logger.info(
+        f"User {user.username} was soft deleted for spam bio: {bio_plus_websites}\n"
+        f"AI Reasoning: {reasoning}"
+    )
+    return Response(
+        data={
+            "message": "This bio seems to be spam. Please contact support@metaculus.com if you believe this was a mistake.",
+            "error_code": "SPAM_DETECTED",
+        },
+        status=status.HTTP_403_FORBIDDEN,
+    )
+
+
+def _update_user_profile(
+    user: User, serializer: UserUpdateProfileSerializer
+) -> Response:
     unsubscribe_tags = serializer.validated_data.get("unsubscribed_mailing_tags")
     if unsubscribe_tags is not None:
         user_unsubscribe_tags(user, unsubscribe_tags)
-
     serializer.save()
-    user.save()
-
     return Response(UserPrivateSerializer(user).data)
 
 
+async def _ask_gpt_to_check_for_spam(
+    bio_plus_websites: str, email: str
```

Comment:
Could you move all these business-logic functions to utils/services not to overload `views.py`?